discovery_engine_core/core/src/engine.rs

```diff
@@ -297,15 +303,26 @@ where
         let mut stacks = self.stacks.write().await;
         let documents =
             SelectionIter::new(BetaSampler, stacks.values_mut()).select(max_documents)?;
+
+        let request_new = if self.request_after < self.core_config.request_after {
+            self.core_config.request_new
+        } else {
+            usize::MAX
+        };
         update_stacks(
             stacks.values_mut(),
             &mut self.ranker,
             history,
             self.core_config.select_top,
             self.core_config.keep_top,
-            self.core_config.request_new,
+            request_new,
         )
         .await?;
+        if self.request_after < self.core_config.request_after {
+            self.request_after += 1;
+        } else {
+            self.request_after = 0;
+        }
```

Comment:
```suggestion
        self.request_after += 1;
        self.request_after %= self.core_config.request_after;
```
more concise, not sure about readability.

Reply:
or even
```rust
self.request_after = (self.request_after + 1) % self.core.config.request_after
```